aggregator_core/src/datastore.rs

```diff
@@ -3181,18 +3211,27 @@ impl<C: Clock> Transaction<'_, C> {
     ) -> Result<Option<BatchAggregation<SEED_SIZE, Q, A>>, Error> {
         let stmt = self
             .prepare_cached(
-                "SELECT
-                    batch_aggregations.state, aggregate_share, report_count, checksum
+                "WITH non_gc_batches AS (
+                    SELECT
+                        tasks.task_id, batch_identifier, aggregation_param
+                    FROM batch_aggregations
+                    JOIN tasks ON tasks.id = batch_aggregations.task_id
+                    WHERE tasks.task_id = $1
+                      AND batch_aggregations.batch_identifier = $2
+                      AND batch_aggregations.aggregation_param = $3
+                    GROUP BY tasks.task_id, batch_identifier, aggregation_param
+                    HAVING MAX(UPPER(COALESCE(batch_interval, client_timestamp_interval))) >= COALESCE($5::TIMESTAMP - MAX(tasks.report_expiry_age) * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)
+                )
+                SELECT
+                    client_timestamp_interval, batch_aggregations.state, aggregate_share,
+                    report_count, checksum, aggregation_jobs_created, aggregation_jobs_terminated
                 FROM batch_aggregations
                 JOIN tasks ON tasks.id = batch_aggregations.task_id
-                JOIN batches ON batches.task_id = batch_aggregations.task_id
-                            AND batches.batch_identifier = batch_aggregations.batch_identifier
-                            AND batches.aggregation_param = batch_aggregations.aggregation_param
                 WHERE tasks.task_id = $1
                   AND batch_aggregations.batch_identifier = $2
                   AND batch_aggregations.aggregation_param = $3
                   AND ord = $4
-                  AND UPPER(COALESCE(batches.batch_interval, batches.client_timestamp_interval)) >= COALESCE($5::TIMESTAMP - tasks.report_expiry_age * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)",
+                  AND EXISTS(SELECT 1 FROM non_gc_batches WHERE task_id = $1 AND batch_identifier = $2 AND aggregation_param = $3)",
```

Comment:
1. I took a crack at a comment, let me know if this makes sense. [Somewhat off-topic, but it's unfortuante that we call a "batch" something identified by (task ID, batch identifier, aggregation param) -- ideally, a batch would be identified by (task ID, batch identifier). But we don't have a good name for the entity addressed by (task ID, batch identifier, aggregation param).]

2. Surprisingly, it appears that joining against `non_gc_batches` directly is *worse* -- see query plans at end of this comment. It appears that the join version does a sequential scan over `batch_aggregations` for every row with the desired `ord`, then filters that based on a scan through `non_gc_batches`. The original query, OTOH, does a direct index lookup based on the expected values, followed by a filter based on the subquery. This might be affected by the scale of the DB -- I was working against a DB that had run a single `batch_aggregations` unit test, so there were a single-digit number of rows. But for now, I'm OK moving forward with this query.

<details>
<summary>Current query & query plan</summary>

```
EXPLAIN WITH non_gc_batches AS (
                    SELECT
                        tasks.task_id, batch_identifier, aggregation_param
                    FROM batch_aggregations
                    JOIN tasks ON tasks.id = batch_aggregations.task_id
                    WHERE tasks.task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1'
                      AND batch_aggregations.batch_identifier = '\x00000000000003e80000000000000064'
                      AND batch_aggregations.aggregation_param = '\x0c'
                    GROUP BY tasks.task_id, batch_identifier, aggregation_param
                    HAVING MAX(UPPER(COALESCE(batch_interval, client_timestamp_interval))) >= COALESCE(now() - MAX(tasks.report_expiry_age) * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)
                )
                SELECT
                    client_timestamp_interval, batch_aggregations.state, aggregate_share,
                    report_count, checksum, aggregation_jobs_created, aggregation_jobs_terminated
                FROM batch_aggregations
                JOIN tasks ON tasks.id = batch_aggregations.task_id
                WHERE tasks.task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1'
                  AND batch_aggregations.batch_identifier = '\x00000000000003e80000000000000064'
                  AND batch_aggregations.aggregation_param = '\x0c'
                  AND ord = 4
                  AND EXISTS(SELECT 1 FROM non_gc_batches WHERE task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1' AND batch_identifier = '\x00000000000003e80000000000000064' AND aggregation_param = '\x0c')

                                                                                                                                                  QUERY PLAN                                                                                                                                                  
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Result  (cost=16.68..32.73 rows=1 width=124)
   One-Time Filter: $1
   InitPlan 1 (returns $1)
     ->  Subquery Scan on non_gc_batches  (cost=0.29..16.39 rows=1 width=0)
           ->  GroupAggregate  (cost=0.29..16.38 rows=1 width=96)
                 Group Key: tasks_1.task_id, batch_aggregations_1.batch_identifier, batch_aggregations_1.aggregation_param
                 Filter: (max(upper(COALESCE(batch_aggregations_1.batch_interval, batch_aggregations_1.client_timestamp_interval))) >= COALESCE((now() - ((max(tasks_1.report_expiry_age))::double precision * '00:00:01'::interval)), ('-infinity'::timestamp without time zone)::timestamp with time zone))
                 ->  Nested Loop  (cost=0.29..16.34 rows=1 width=168)
                       ->  Index Scan using task_id_index on tasks tasks_1  (cost=0.14..8.16 rows=1 width=48)
                             Index Cond: (task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1'::bytea)
                       ->  Index Scan using batch_aggregations_unique_task_id_batch_id_aggregation_param on batch_aggregations batch_aggregations_1  (cost=0.14..8.17 rows=1 width=136)
                             Index Cond: ((task_id = tasks_1.id) AND (batch_identifier = '\x00000000000003e80000000000000064'::bytea) AND (aggregation_param = '\x0c'::bytea))
   ->  Nested Loop  (cost=16.68..32.73 rows=1 width=124)
         ->  Index Scan using task_id_index on tasks  (cost=0.14..8.16 rows=1 width=8)
               Index Cond: (task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1'::bytea)
         ->  Index Scan using batch_aggregations_unique_task_id_batch_id_aggregation_param on batch_aggregations  (cost=0.14..8.17 rows=1 width=132)
               Index Cond: ((task_id = tasks.id) AND (batch_identifier = '\x00000000000003e80000000000000064'::bytea) AND (aggregation_param = '\x0c'::bytea) AND (ord = 4))
(17 rows)

```
</details>

<details>
<summary>Updated query & query plan</summary>

```
EXPLAIN WITH non_gc_batches AS (
                    SELECT
                        tasks.id AS task_id, batch_identifier, aggregation_param
                    FROM batch_aggregations
                    JOIN tasks ON tasks.id = batch_aggregations.task_id
                    WHERE tasks.task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1'
                      AND batch_aggregations.batch_identifier = '\x00000000000003e80000000000000064'
                      AND batch_aggregations.aggregation_param = '\x0c'
                    GROUP BY tasks.id, batch_identifier, aggregation_param
                    HAVING MAX(UPPER(COALESCE(batch_interval, client_timestamp_interval))) >= COALESCE(now() - MAX(tasks.report_expiry_age) * '1 second'::INTERVAL, '-infinity'::TIMESTAMP)
                )
                SELECT
                    client_timestamp_interval, batch_aggregations.state, aggregate_share,
                    report_count, checksum, aggregation_jobs_created, aggregation_jobs_terminated
                FROM batch_aggregations
JOIN non_gc_batches
ON non_gc_batches.task_id = batch_aggregations.task_id
AND non_gc_batches.batch_identifier = batch_aggregations.batch_identifier
AND non_gc_batches.aggregation_param = batch_aggregations.aggregation_param
AND ord = 4;

                                                                                                                                             QUERY PLAN                                                                                                                                             
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 Nested Loop  (cost=16.35..29.55 rows=1 width=124)
   Join Filter: ((batch_aggregations.task_id = tasks.id) AND (batch_aggregations.batch_identifier = batch_aggregations_1.batch_identifier) AND (batch_aggregations.aggregation_param = batch_aggregations_1.aggregation_param))
   ->  Seq Scan on batch_aggregations  (cost=0.00..13.12 rows=1 width=196)
         Filter: (ord = 4)
   ->  GroupAggregate  (cost=16.35..16.40 rows=1 width=72)
         Group Key: tasks.id, batch_aggregations_1.batch_identifier, batch_aggregations_1.aggregation_param
         Filter: (max(upper(COALESCE(batch_aggregations_1.batch_interval, batch_aggregations_1.client_timestamp_interval))) >= COALESCE((now() - ((max(tasks.report_expiry_age))::double precision * '00:00:01'::interval)), ('-infinity'::timestamp without time zone)::timestamp with time zone))
         ->  Sort  (cost=16.35..16.36 rows=1 width=144)
               Sort Key: tasks.id
               ->  Nested Loop  (cost=0.29..16.34 rows=1 width=144)
                     ->  Index Scan using task_id_index on tasks  (cost=0.14..8.16 rows=1 width=16)
                           Index Cond: (task_id = '\xaf6d8d803ec6d78832c8fcb420f85bbe7b24c5ae4a540d0525849df9127cfdc1'::bytea)
                     ->  Index Scan using batch_aggregations_unique_task_id_batch_id_aggregation_param on batch_aggregations batch_aggregations_1  (cost=0.14..8.17 rows=1 width=136)
                           Index Cond: ((task_id = tasks.id) AND (batch_identifier = '\x00000000000003e80000000000000064'::bytea) AND (aggregation_param = '\x0c'::bytea))
(14 rows)
```
</details>

3. The purpose of `non_gc_batches` is to evaluate the GC condition, which requires evaluating all shards of `batch_aggregations`; thus, grouping (across shards) is necessary in that query. But this means we can't return `batch_aggregations.id`, since we aren't grouping (& can't group) over it; a query with a `GROUP BY` clause can only return group-by'ed columns or aggregate columns: https://www.postgresql.org/docs/current/sql-select.html#SQL-GROUPBY

Reply:
Thanks for taking the trouble of trying out the JOIN. That result is indeed surprising, but good enough justification for structuring the query as you already did. I think an explanatory comment will suffice.